Make *_range functions consistent

jschendel opened the pull request:
- [X] closes #17471
- [X] tests added / passed
- [X] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [X] whatsnew entry

Comments:

- I couldn't find any existing tests of `period_range`, so I created a new file `test_period_range.py` in a similar fashion to the existing files `test_date_range.py` and `test_timedelta_range.py`.
- Cleaned up a few existing tests relevant to this issue (used `with` context where appropriate, etc.).

jschendel committed:
Make *_range functions consistent

gfyoung commented on doc/source/whatsnew/v0.21.0.txt:
```diff
@@ -310,6 +310,59 @@ Previously, :func:`to_datetime` did not localize datetime ``Series`` data when `
 
 Additionally, DataFrames with datetime columns that were parsed by :func:`read_sql_table` and :func:`read_sql_query` will also be localized to UTC only if the original SQL columns were timezone aware datetime columns.
 
+.. _whatsnew_0200.api.consistency_of_range_functions:
```
You mean "0210"

gfyoung commented on pandas/core/indexes/datetimes.py:
```diff
@@ -2030,7 +2031,7 @@ def date_range(start=None, end=None, periods=None, freq='D', tz=None,
     Notes
     -----
-    2 of start, end, or periods must be specified
+    Exactly two of start, end, or periods must be specified
```
I know what you're trying to say here, but it's a little confusing the final parameter is plural.  Say something like this (and propagate to places where relevant):

"Of the three parameters, start, end, and periods, exactly two must be specified."

For the doc-strings, I would even consider ticking them for additional clarity.

gfyoung commented on pandas/tests/indexes/datetimes/test_date_range.py:
```diff
@@ -146,14 +146,26 @@ def test_date_range_businesshour(self):
 
     def test_range_misspecified(self):
         # GH #1095
+        with pytest.raises(ValueError):
```
We know what the error message is here (you just added it).  Let's make this test (and all other relevant ones) stronger by checking the error message too.

jschendel committed:
gfyoung review

Clean error message, check for error in tests, fix typo in whatsnew.

jreback commented on doc/source/whatsnew/v0.21.0.txt:
```diff
@@ -310,6 +310,59 @@ Previously, :func:`to_datetime` did not localize datetime ``Series`` data when `
+  ---------------------------------------------------------------------------
+  ValueError: Of the three parameters, start, end, and periods, exactly two must be specified
+
+Additionally, the endpoint parameter ``end`` was not included in the intervals produced by ``interval_range``.  However, all other range functions include ``end`` in their output.  To promote consistency among the range functions, ``interval_range`` will now include ``end`` as the right endpoint of the final interval, except if ``freq`` is specified in a way which skips ``end``.
```
is this last sentence true for period as well?

jschendel replied:
I don't think `periods` can cause a skip.  Since we're forcing exactly two of `start`/ `end`/`periods` to be passed, if `periods` is passed then only one of `start`/`end` can be passed.  The _range functions then compute the non-passed using `periods`/ `freq`/passed, so by construction nothing can get skipped.  I think the only case a skip is possible is `end` being skipped when `start`/`end`/`freq` are all passed.

jreback commented on pandas/core/indexes/datetimes.py:
```diff
@@ -412,7 +412,8 @@ def __new__(cls, data=None,
                   tz=None, normalize=False, ambiguous='raise', closed=None):
         if com._count_not_none(start, end, periods) != 2:
-            raise ValueError('Must specify two of start, end, or periods')
+            raise ValueError('Of the three parameters, start, end, and '
```
``Of the three parameters: start, end, and periods, `` comma -> ``:``

jreback commented on pandas/core/indexes/datetimes.py:
```diff
@@ -2030,7 +2031,8 @@ def date_range(start=None, end=None, periods=None, freq='D', tz=None,
     Notes
     -----
-    2 of start, end, or periods must be specified
+    Of the three parameters, ``start``, ``end``, and ``periods``, exactly two
```
same (and all other occurrences)

jreback commented on pandas/core/indexes/interval.py:
```diff
@@ -1039,43 +1039,37 @@ def interval_range(start=None, end=None, freq=None, periods=None,
         Right bound for generating data
-    freq : interger, string or DateOffset, default 1
-    periods : interger, default None
+    freq : integer, string or DateOffset, default 1
```
somewhere mention (if not already) that freq must be consistent with the type of the start/end

jreback replied:
also I *think* there are tests but for this, but pls confirm.

jschendel replied:
Confirmed that there are tests for mixed units in `TestIntervalRange.test_errors`.  Adding one additional test to be safe.

jreback commented on pandas/tests/indexes/period/test_period_range.py:
```diff
@@ -0,0 +1,54 @@
+            period_range(periods=5)
+
+        with tm.assert_raises_regex(ValueError, msg):
+            period_range()
```
add inconsistent freq w.r.t. start/end

jschendel replied:
The current behavior on 0.20.3 (and this PR as-is) is not to raise, but to convert to `freq`:

```python
In [2]: p1 = pd.Period('2017-01-01', freq='W')
   ...: p2 = pd.Period('2017-12-31', freq='W')
   ...: pd.period_range(start=p1, end=p2, freq='Q')
   ...:
Out[2]: PeriodIndex(['2017Q1', '2017Q2', '2017Q3', '2017Q4'], dtype='period[Q-DEC]', freq='Q-DEC')
```

When `freq` isn't specified, it converts to daily (default is `freq='D'`):
```python
In [5]: pd.period_range(start=p1, end=p2)
Out[5]:
PeriodIndex(['2017-01-01', '2017-01-02', '2017-01-03', '2017-01-04',
             '2017-01-05', '2017-01-06', '2017-01-07', '2017-01-08',
             '2017-01-09', '2017-01-10',
             ...
             '2017-12-22', '2017-12-23', '2017-12-24', '2017-12-25',
             '2017-12-26', '2017-12-27', '2017-12-28', '2017-12-29',
             '2017-12-30', '2017-12-31'],
            dtype='period[D]', length=365, freq='D')
```

Same behavior with different frequencies for all of `start`/`end`/`freq`:

```python
In [4]: p3 = pd.Period('2017-12-31', freq='M')
   ...: pd.period_range(start=p1, end=p3, freq='Q')
   ...:
Out[4]: PeriodIndex(['2017Q1', '2017Q2', '2017Q3', '2017Q4'], dtype='period[Q-DEC]', freq='Q-DEC')
```

Should this raise if any of `start`/`end`/`freq` are passed with different frequencies?  If `start`/`end` are passed with the same frequency, and `freq` isn't specified, should `freq` be set to match `start`/`end` instead of using the daily default?

Couldn't find a direct comparison with the other range functions to check against for consistency.  `interval_range` will raise if `freq` isn't compatible with they type of `start`/`end`, but that seems a bit different.

jreback replied:
I guess this is meant to be the 'new' freq, while the start/end serve as the anchor endpoints. this seems fine then. (maybe adding the above example to the docs would be helpful though).

jschendel replied:
Okay, will add tests for this behavior.

jreback reviewed (changes requested):
small comments. looks really good. a couple of in-line questions. ping when pushed.

jschendel committed:
jreback review and misc changes

addressed jreback's review, cleaned up docstrings, added documention, update to interval_range endpoint fix, additional tests

jschendel commented:
ping @jreback

Made the review changes, as well as a few additional things I noticed:
- Found and fixed a corner case bug in my endpoint fix for `interval_range`, which resulted in shuffling around some of that code
  - Also added some additional checks around `periods` (similar to what's done in the other functions)
  - Also added some additonal tests
- Regarding `period_range` when `start` or `end` are `Period` objects (discussed in review comments):
   - Added tests to for this scenario
   - Added an Example section to the `period_range` docstring covering this scenario
   - Added an example of this using the `PeriodIndex` constructor in `timeseries.rst`
- Cleaned up the docstrings of each function and made them as consistent with each other as possible
- Cleaned up a few error messages to make the capitalization consistent with param names

I also noticed string/datetime-like `start`/`end` doesn't work with `interval_range`, despite the docstring saying it does (don't think it ever did).  Should that be fixed here, or is a separate issue/PR more appropriate?

jreback commented on doc/source/whatsnew/v0.21.0.txt:
```diff
@@ -311,6 +311,59 @@ Previously, :func:`to_datetime` did not localize datetime ``Series`` data when `
+
+  In [2]: pd.interval_range(start=0, end=4, periods=6)
+  ---------------------------------------------------------------------------
+  ValueError: Of the three parameters, start, end, and periods, exactly two must be specified
```
use : after parameters here

jreback commented on pandas/core/indexes/datetimes.py:
```diff
@@ -2013,24 +2014,25 @@ def date_range(start=None, end=None, periods=None, freq='D', tz=None,
-    periods : integer or None, default None
-        If None, must specify start and end
+    periods : integer, default None
+        Number of dates to generate
```
number of periods

jreback commented on pandas/core/indexes/datetimes.py:
```diff
@@ -2056,24 +2058,25 @@ def bdate_range(start=None, end=None, periods=None, freq='B', tz=None,
+        Number of dates to generate
     freq : string or DateOffset, default 'B' (business daily)
-        Frequency strings can have multiples, e.g. '5H'
+        Frequency strings can have multiples, e.g. '5, default
```
the example here looks wrong

jreback commented on pandas/core/indexes/interval.py:
```diff
@@ -1028,54 +1029,61 @@ def func(self, other):
+                         'exactly two must be specified')
+
+    # must all be same units or None
+    arr = np.array(list(com._not_none(start, end, freq)))
```
you could just do this above and compare ``len(arr) != 2``

jreback commented on pandas/core/indexes/interval.py:
```diff
@@ -1028,54 +1029,61 @@ def func(self, other):
+        The length of each interval. Must be consistent with the
+        type of start and end
+    name : string, default None
+        Name of the resulting IntervalIndex
```
can you add a couple of examples?

jreback commented on pandas/core/indexes/timedeltas.py:
```diff
@@ -960,22 +961,22 @@ def _generate_regular_range(start, end, periods, offset):
+        Name of the resulting TimedeltaIndex
+    closed : string, default None
         Make the interval closed with respect to the given frequency to
         the 'left', 'right', or both sides (None)
```
can you add some examples

jreback reviewed (changes requested):
couple of minor comments.

jreback commented:
these are all fine: https://github.com/pandas-dev/pandas/pull/17482#issuecomment-328436966 thanks!

> I also noticed string/datetime-like start/end doesn't work with interval_range, despite the docstring saying it does (don't think it ever did). Should that be fixed here, or is a separate issue/PR more appropriate

Datetimelike *should* work with ``interval_range``, as it works with the regular constructors. I don't think strings should work though, its tricky to infer what is mean (IOW it could be a datetime, or a timedelta, or it could be an error). I would raise on strings (but accept datetimelike)

jreback commented:
ping on push / green.

jschendel committed:
Review and datetime-like implementation

Made review related changes, implemented support for datetime-like input in interval_range.

jschendel commented:
ping @jreback 

Made the review changes, and additionally:

- Implemented support for datetime-like in `interval_range`
   - Added a bunch of associated tests
   - Related to previous review comment: note that both not null checks are actually necessary, as one is checking with `periods`, the other `freq`
- Changed a couple `ValueError` to `TypeError`, as that seems like the more appropriate exception

jreback commented on pandas/core/indexes/interval.py:
```diff
@@ -1028,54 +1033,137 @@ def func(self, other):
-            raise ValueError("must specify 2 of start, end, periods")
+    if is_number(start):
+        is_datetime_interval = False
+    elif start is not None:
```
This not correct for the start is None (and end). these *must* be number, datetime, or timedelta. For datetime, timedelta, then you can wrap the Timestamp/Timedelta around.

jschendel committed:
fix datetimelike

jschendel commented:
ping @jreback 

Updated datetime-like support.  For some reason I originally thought that datetime-like didn't include timedelta, so I shortcut to just parsing timestamps.  Updated with support for timedelta and added timedelta related tests.

jreback commented on pandas/core/indexes/datetimes.py:
```diff
@@ -2091,7 +2094,7 @@ def bdate_range(start=None, end=None, periods=None, freq='B', tz=None,
                 normalize=True, name=None, closed=None, **kwargs):
     """
-    **EXPERIMENTAL** Return a fixed frequency datetime index, with
+    **EXPERIMENTAL** Return a fixed frequency DatetimeIndex, with
```
side issue we can prob remove the experimental (but new PR for that)

jschendel replied:
Did this in #17554, as there were other doc related fixes that needed to be made.

jreback commented on pandas/core/indexes/interval.py:
```diff
@@ -1028,54 +1033,147 @@ def func(self, other):
+
+    iv_type = {'numeric': True, 'timestamp': True, 'timedelta': True}
+
+    start = com._maybe_box_datetimelike(start)
```
how about a helper function to validate this instead of repeating code.
maybe
```
if is_number(start) or is_number(end):
    boxer = type(start) if start is not None else type(end)
elif isinstance(start, Timestamp) or isinstance(end, Timestamp):
   boxer = Timestamp
elif .....
    ....
else:
    raise
```

jreback reviewed (changes requested):
see if you can condense the constructor a bit. yeah I know a lot is going on,.

jschendel committed:
clean interval_range

jschendel commented:
@jreback : cleaned up the `interval_range` constructor

jreback commented on doc/source/whatsnew/v0.21.0.txt:
```diff
@@ -358,6 +358,59 @@ Previously, :func:`to_datetime` did not localize datetime ``Series`` data when `
+Consistency of Range Functions
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+In previous versions, there were some inconsistencies between the various range functions: ``date_range``, ``bdate_range``, ``cdate_range``, ``interval_range``, ``period_range``, and ``timedelta_range``.   (:issue:`17471`).
```
can you add interval_range to the api.rst (might as well add cdate_range too), can you add these as `` :func:`date_range` `` as well.

jreback commented on pandas/core/indexes/interval.py:
```diff
@@ -1028,54 +1034,152 @@ def func(self, other):
+        msg = 'end must be numeric or datetime-like, got {end}'
+        raise ValueError(msg.format(end=end))
+
+    if is_float(periods):
```
much cleaner, nice!

jreback commented on pandas/core/indexes/interval.py:
```diff
@@ -1028,54 +1034,152 @@ def func(self, other):
-            raise ValueError("must specify 2 of start, end, periods")
+    if is_number(endpoint):
+        if periods is None:
+            periods = int((end - start) // freq)
```
is this what we do elsewhere? this is effectively floored not that this is bad, just wondering if its consistent

jschendel replied:
It's floored elsewhere too, e.g. https://github.com/pandas-dev/pandas/blob/f11bbf2f505d81900cc83ce387a6a1b1d2a2f866/pandas/core/indexes/period.py#L199-L200

https://github.com/pandas-dev/pandas/blob/fa557f7391589f351b1260f46b3b3db22492f50b/pandas/core/indexes/datetimes.py#L292-L293

jreback reviewed (approved):
looking really good. just a couple of minor doc changes and a question

jschendel committed:
minor doc changes

jschendel commented:
@jreback : Made the doc changes.  Didn't see an obvious location for `interval_range` in api.rst, so I created a new section.  Additionally fixed one minor typo in whatsnew.

jreback commented:
thanks @jschendel nice PR, and very responsive! keep em' comin!

jorisvandenbossche commented on doc/source/api.rst:
```diff
@@ -218,10 +218,19 @@ Top-level dealing with datetimelike
    to_timedelta
    date_range
    bdate_range
+   cdate_range
```
this one is not in the public API, so adding it here does not work

jschendel replied:
Thanks, see #17554

jorisvandenbossche commented on doc/source/whatsnew/v0.21.0.txt:
```diff
@@ -358,6 +358,59 @@ Previously, :func:`to_datetime` did not localize datetime ``Series`` data when `
+Consistency of Range Functions
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+In previous versions, there were some inconsistencies between the various range functions: func:`date_range`, func:`bdate_range`, func:`cdate_range`, func:`period_range`, func:`timedelta_range`, and func:`interval_range`. (:issue:`17471`).
```
You are missing `:`'s for all references: `:func:..` instead of `func:..`

jorisvandenbossche replied:
(and also cdate_range here does not work)

jschendel replied:
Thanks, see #17554
